File: frontend/src/types/environments.ts
export interface PostgresEnvironment {
  id: string;
  project_id: string;
  name: string;
  host: string;
  port: number;
  database: string;
  username: string;
  password: string;
  container_name?: string | null;
  volume_name?: string | null;
  status: string;
  created_at?: string | null;
  managed: boolean;
}

export interface CreateEnvironmentPayload {
  name: string;
  project_id: string;
}

export interface EnvironmentsResponse {
  environments: PostgresEnvironment[];
  active_environment_id: string | null;
}

export interface ActiveEnvironmentResponse {
  environment: PostgresEnvironment | null;
  stable_endpoint: string;
}

export interface SqlExecutionResponse {
  columns: string[];
  rows: Record<string, unknown>[];
  row_count: number;
  command: string;
}

File: backend/app/schemas/environments.py
from pydantic import BaseModel, Field


class EnvironmentCreate(BaseModel):
    name: str = Field(min_length=1)
    host: str = Field(min_length=1)
    port: int = Field(default=5432, ge=1, le=65535)
    database: str = Field(min_length=1)
    username: str = Field(min_length=1)
    password: str = Field(min_length=1)
    project_id: str = "default"


class Environment(EnvironmentCreate):
    id: str
    container_name: str | None = None
    volume_name: str | None = None
    status: str = "external"
    created_at: str | None = None
    managed: bool = False


class EnvironmentProvisionRequest(BaseModel):
    name: str = Field(min_length=1)
    project_id: str = "default"


class SqlExecutionRequest(BaseModel):
    sql: str = Field(min_length=1)


class SqlExecutionResponse(BaseModel):
    columns: list[str]
    rows: list[dict]
    row_count: int
    command: str


class EnvironmentListResponse(BaseModel):
    environments: list[Environment]
    active_environment_id: str | None


class ActiveEnvironmentResponse(BaseModel):
    environment: Environment | None
    stable_endpoint: str


class SwitchActiveResponse(BaseModel):
    active: Environment
    stable_endpoint: str
